### PR TITLE
Replace bool on Decoder trait with the DecoderStatus type (includes Encoder side too)

### LINKED ISSUE #6019: For consideration: replace `bool` in encoding with something semantically clearer
The rust guidelines recommend not using `bool` where it's unclear what means what, which I think is the case of return values in `Encoder` an `Decoder`. I also had to to double check I remember the meaning a few times.

Before stabilization it'd be nice to think through our options:

* keep `bool` - it's easy to write, just harder to read
* use a custom enum - pros: we can define semantics, names etc as we wish and in a few thousand years we can define `?` operator as we wish
* use `ControlFlow<(), ()>` - this is already provided but has bunch of issues - see below
* use `Poll<()>` - could cause confusion in `Future` implementations because `Poll::Pending` from a decoder is unrelated to `Poll::Pending` returned from IO - the latter must be returned from `Future`, not the former. Also its `?` impl sucks becaue it only works with `Result` which we don't always need

The issues with `ControlFlow`:

* long name with weird generics (we could typedef but see below)
* the behavior of `?` is inverse of what we would like: `Continue` suggests "continue decoding, it's not done yet", `Break` suggests "decoded" but `Break` returns from a function, `Continue` does not, which is opposite of what we want in decoder impls.
* once we lock in to use it we can't ever fix `?`

It seems to me that custom enum is the least bad option:

```rust
/// Indicates whether an encoder or decoder reached its final state.
#[must_use]
pub enum Status<T = ()> {
    // Encoder has more data to encode or decoder needs more data to decode
    PendingData,
    // Encoder or decoder finished encoding, in case of decoder, this contains Result which needs to be checked and `end` needs to be called in case it's Ok.
    Done(T),
}

impl Status {
    fn is_pending_data(&self) -> bool { matches!(self, Status::PendingData) }
    fn is_done(&self) -> bool { matches!(self, Status::Done(_)) }
}
```

I could be also convinced to have separate enums for `Status`.

### PR DESCRIPTION
**Pull request by nyonson**
Builds on #6164 

Bit of a race to the finish and tons of little changes, so I will make a few more passes. But figured Kixunil could take the lint/mutant/doc fixes if he wants, or give thoughts on the `DecoderStatus` names.

Closes #6019

### CODE REVIEW

File: consensus_encoding/tests/compact_size.rs [hunk 1]
```diff
@@ -307,3 +307,21 @@ fn decoder_compact_size_end_incomplete_nine_byte() {
...
+fn encoder_compact_size_len_matches_encoded_bytes() {
+    // Test that CompactSizeEncoder::len() returns the correct length for various values
+    let test_cases = vec![
+        (0usize, 1),              // Single byte: 0x00
+        (252usize, 1),            // Single byte: 0xFC
+        (253usize, 3),            // Three bytes: 0xFD + 2 bytes
+        (0xFFFF_usize, 3),        // Three bytes: 0xFD + 2 bytes
+        (0x1_0000_usize, 5),      // Five bytes: 0xFE + 4 bytes
+        (0xFFFF_FFFF_usize, 5),   // Five bytes: 0xFE + 4 bytes
+    ];
```
Comment (Kixunil) [inline consensus_encoding/tests/compact_size.rs:355]: Just to avoid FPs in linters and whatnot, I'd remove the `vec!` part, leaving in a simple array.

**Review by Kixunil — COMMENTED** [review]
Cool, thanks a lot for the help!

File: consensus_encoding/src/decode/decoders.rs [hunk 2]
```diff
@@ -111,8 +112,11 @@ impl Decoder for ByteVecDecoder {
         self.bytes_written += copy_len;
         *bytes = &bytes[copy_len..];
 
-        // Return true if we still need more data.
-        Ok(self.bytes_written < self.bytes_expected)
+        if self.bytes_written < self.bytes_expected {
+            Ok(DecoderStatus::NeedsMore)
+        } else {
+            Ok(DecoderStatus::Ready)
```
Comment (Kixunil) [inline consensus_encoding/src/decode/decoders.rs:118]: FTR, I like the names.

**Review by Kixunil — COMMENTED** [review]
I noticed there is a similar thing with tests as in enocding implementation: there are tests that simply call `push_bytes` but I think we should have something like `check_decode_from_slice<T: Decodable + Eq>(bytes: &[u8], expected: &T)` which decodes the value using two strategies:

* entire slice fed (until it's consumed)
* feeding one byte at a time

This should cover most edge cases, I think. Not necessary for the PR but FYI, the reason I did such in mine was to reduce the total number of changes (since I then didn't have to change tons of tests again).

**Comment by nyonson** [comment]
> I noticed there is a similar thing with tests as in enocding implementation: there are tests that simply call `push_bytes` but I think we should have something like `check_decode_from_slice<T: Decodable + Eq>(bytes: &[u8], expected: &T)`

I'd like to try an mirror the encoder `check_*`s just for API balance. If these are just for tests I don't think that should be too hard?

**Comment by nyonson** [comment]
[7002394](https://github.com/rust-bitcoin/rust-bitcoin/commit/7002394b9fae0ead008a3add54a83da4c309b1af): fix test import merge conflicts

**Comment by nyonson** [comment]
At this point we have:

* `EncoderStatus::has_more()` and `EncoderStatus::has_finished()`
* `DecoderStatus::needs_more()` and `DecoderStatus::is_ready()`

I like how they are symmetrical. Have a slight preference to change `EncoderStatus::has_finished()` to `EncoderStatus::is_finished()`. I tried to think of more idiomatic (`is_*`) versions of `has_more` and `needs_more` and they just felt very clunky, but open to suggestions.

File: consensus_encoding/src/decode/mod.rs [hunk 3]
```diff
@@ -58,23 +58,23 @@ pub trait Decoder: Sized {
 
     /// Pushes bytes into the decoder, consuming as much as possible.
     ///
-    /// The slice reference will be advanced to point to the unconsumed portion. Returns `Ok(true)`
-    /// if more bytes are needed to complete decoding, `Ok(false)` if the decoder is ready to
+    /// The slice reference will be advanced to point to the unconsumed portion. Returns `Ok(DecoderStatus::NeedsMore)`
+    /// if more bytes are needed to complete decoding, `Ok(DecoderStatus::Ready)` if the decoder is ready to
```
Comment (tcharding) [inline consensus_encoding/src/decode/mod.rs]: Can we wrap the line at 100 characters pls, 1.0 is chasing perfect.

File: consensus_encoding/src/decode/mod.rs [hunk 4]
```diff
@@ -58,23 +58,23 @@ pub trait Decoder: Sized {
...
     /// finalize with [`Self::end`], or `Err(error)` if parsing failed.
     ///
     /// # Errors
     ///
     /// Returns an error if the provided bytes are invalid or malformed according to the decoder's
     /// validation rules. Insufficient data (needing more bytes) is *not* an error for this method,
-    /// the decoder will simply consume what it can and return `true` to indicate more data is
+    /// the decoder will simply consume what it can and return `DecoderStatus::NeedsMore` to indicate more data is
```
Comment (tcharding) [inline consensus_encoding/src/decode/mod.rs]: Wrap here too.

File: consensus_encoding/src/decode/decoders.rs (hunk 2 above)
Comment (tcharding) [reply consensus_encoding/src/decode/decoders.rs:118]: +1

File: consensus_encoding/src/decode/mod.rs [hunk 5]
```diff
@@ -100,6 +100,35 @@ pub trait Decoder: Sized {
...
 
+/// Indicates whether a decoder needs more data or is ready to finalize.
+///
+/// This is returned from the [`Decoder::push_bytes`] method to indicate whether the decoder
+/// should continue accumulating data or is ready to produce the decoded value with [`Decoder::end`].
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum DecoderStatus {
+    /// The decoder needs more data to complete decoding.
+    ///
+    /// Continue pushing byte slices with [`Decoder::push_bytes`] until this status changes to [`Ready`](DecoderStatus::Ready).
```
Comment (tcharding) [inline consensus_encoding/src/decode/mod.rs]:
```suggestion
    /// Continue pushing byte slices with [`Decoder::push_bytes`] until this status changes to
    /// [`Ready`](DecoderStatus::Ready).
```

File: consensus_encoding/tests/composition.rs [hunk 6]
```diff
@@ -46,16 +58,18 @@ impl core::fmt::Display for CompositeError {
...
     inner: Decoder2<ArrayDecoder<4>, ArrayDecoder<2>>,
 }
 
+#[cfg(feature = "alloc")]
 impl Decoder for CompositeDataDecoder {
     type Output = CompositeData;
     type Error = CompositeError;
 
-    fn push_bytes(&mut self, bytes: &mut &[u8]) -> Result<bool, Self::Error> {
+    fn push_bytes(&mut self, bytes: &mut &[u8]) -> Result<bitcoin_consensus_encoding::DecoderStatus, Self::Error> {
```
Comment (tcharding) [inline consensus_encoding/tests/composition.rs]: Can we import the type (in this test module only, in the other places the path is good).

File: consensus_encoding/src/decode/mod.rs (hunk 3 above)
Comment (tcharding) [reply consensus_encoding/src/decode/mod.rs]: In case you got to wild after reading this; IMO no need to always wrap list items at 100 characters.

**Comment by tcharding** [comment]
I rekon we should have a patch at the end that runs the formatter. It would suck if we released `encoding` with all these formating issues in it.

**Comment by tcharding** [comment]
FWIW this is really nice, top effort @Kixunil and @nyonson

**Comment by tcharding** [comment]
reviewed: 1e4ac33a7e0adfe019da0ca161980480f47ad404

**Comment by nyonson** [comment]
50d347c1417e76b00134e7d87b86ed24b55a7024: wrapped docs at 100 characters, cleaned up the tests import, and added a commit to format `consensus_encoding`.

**Review by tcharding — APPROVED** [review]
ACK 50d347c1417e76b00134e7d87b86ed24b55a7024

**Comment by tcharding** [comment]
LFG!

**Comment by apoelstra** [comment]
In a followup we may want to clarify what happens if you call `Decoder::push_bytes` after getting `Ok(Ready)`. Will you just get `Ok(Ready)` again or is this unspecified behavior?

I won't hold up the PR (or release) over this.

**Comment by apoelstra** [comment]
In 164eed18ab2ea24c31399ac47444f8d14074b104:

This commit (and the following one) don't compile:

```
consensus_encoding% cargo test --no-default-features
    Updating crates.io index
     Locking 1 package to latest compatible version
      Adding bitcoin v0.32.9
   Compiling bitcoin-internals v0.5.0 (/home/apoelstra/code/rust-bitcoin/rust-bitcoin/pr-review/internals)
   Compiling hex-conservative v0.3.2
   Compiling bitcoin-consensus-encoding v0.2.0 (/home/apoelstra/code/rust-bitcoin/rust-bitcoin/pr-review/consensus_encoding)
error[E0432]: unresolved imports `bitcoin_consensus_encoding::drain_to_vec`, `bitcoin_consensus_encoding::encode_to_vec`
  --> consensus_encoding/tests/composition.rs:7:5
   |
 7 |     drain_to_vec, Encode, encode_to_vec, Encoder, Encoder2, Encoder3, Encoder6, UnexpectedEofError,
   |     ^^^^^^^^^^^^          ^^^^^^^^^^^^^ no `encode_to_vec` in the root
   |     |
   |     no `drain_to_vec` in the root
   |
note: found an item that was configured out
  --> /home/apoelstra/code/rust-bitcoin/rust-bitcoin/pr-review/consensus_encoding/src/lib.rs:94:24
   |
92 | #[cfg(feature = "alloc")]
   |       ----------------- the item is gated behind the `alloc` feature
93 | #[doc(inline)]
94 | pub use self::encode::{drain_to_vec, encode_to_vec};
   |                        ^^^^^^^^^^^^
note: found an item that was configured out
  --> /home/apoelstra/code/rust-bitcoin/rust-bitcoin/pr-review/consensus_encoding/src/lib.rs:94:38
   |
92 | #[cfg(feature = "alloc")]
   |       ----------------- the item is gated behind the `alloc` feature
93 | #[doc(inline)]
94 | pub use self::encode::{drain_to_vec, encode_to_vec};
   |                                      ^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0432`.
error: could not compile `bitcoin-consensus-encoding` (test "composition") due to 1 previous error
warning: build failed, waiting for other jobs to finish...
```

**Comment by apoelstra** [comment]
50d347c1417e76b00134e7d87b86ed24b55a7024 needs rebase

**Comment by nyonson** [comment]
9ecae87d710b432d19da328a79721b5c6c52555e: fix up broken commits, add docs on `push_bytes`, and fix up merge conflicts

**Review by tcharding — APPROVED** [review]
ACK 9ecae87d710b432d19da328a79721b5c6c52555e

**Review by apoelstra — APPROVED** [review]
ACK 9ecae87d710b432d19da328a79721b5c6c52555e; successfully ran local tests

File: consensus_encoding/src/decode/decoders.rs (hunk 2 above)
Comment (yancyribbens) [reply consensus_encoding/src/decode/decoders.rs:118]: FTR I think this name is confusing.  Does `Ready` mean that it's complete, or that it hasn't started yet and is ready to start?  It would be less ambiguis to say `DecoderStatus::Complete` or ` DecoderStatus::Done` or `DecoderStatus::Finished`.  Anything to let the reader know the state has moved from NeedsMore to is done.

Comment (yancyribbens) [reply consensus_encoding/src/decode/decoders.rs:118]: Sorry to bikeshed after the fact, but I didn't read this PR in full, and now I'm trying to make heads or tales of whats going on with the new changes.

Comment (nyonson) [reply consensus_encoding/src/decode/decoders.rs:118]: Subjective, but I think `Ready` is better than `Finished` (which we use on the encoder status) because there is still work to be done. The caller need to call `end` now to get the decoded object.

Comment (nyonson) [reply consensus_encoding/src/decode/decoders.rs:118]: Also, these enums won't actually be seen by the end users very often. They will interact with the drivers which wrap them, so I don't think these names matter all that much.

Comment (yancyribbens) [reply consensus_encoding/src/decode/decoders.rs:118]: Well I ended up here because I am reading this line https://github.com/rust-bitcoin/rust-bitcoin/blob/a9504e539d9d7e07934343918060e467b09f8607/p2p/src/message.rs#L1359, and before, it was !need_more which to me was understandable (there is no more bytes to consume).  Although now it says ready, which just makes me think, ready for what lol.

Comment (nyonson) [reply consensus_encoding/src/decode/decoders.rs:118]: Eh, just not sure this is really all that different.

Comment (yancyribbens) [reply consensus_encoding/src/decode/decoders.rs:118]: Can't say I agree in this case.  Although, all I can do is make my point and let others decide.

Comment (yancyribbens) [reply consensus_encoding/src/decode/decoders.rs:118]: hmm, reading this a bit more, `needs_more` and `ready` are mutually exclusive, yet they are defined separately?

```rust
impl DecoderStatus {
    /// Returns `true` if the decoder needs more data to continue.
    pub fn needs_more(&self) -> bool { matches!(self, Self::NeedsMore) }
 
    /// Returns `true` if ready to produce decoded value with [`Decoder::end`].
    pub fn is_ready(&self) -> bool { matches!(self, Self::Ready) }
}
```

One could be defined in terms of the other, or you could just use the negation as before, like `!need_more()`.  Anyway, I guess it is what it is at this point.
